modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java

```diff
@@ -1092,11 +1092,13 @@ private void addAllToBatch(
      * Adds modified entries to the watch event queue.
      */
     private void queueWatchEvent() {
-        if (recoveryStatus.get() == RecoveryStatus.INITIAL) {
-            // Watches haven't been enabled yet, no need to queue any events, they will be replayed upon recovery.
-            updatedEntries.clear();
-        } else {
-            notifyWatchProcessor(updatedEntries.toNotifyWatchProcessorEvent(rev));
+        synchronized (watchProcessorMutex) {
```

Comment:
In general, I'm okay with the patch, but we must be careful in the future in terms of a potential deadlocks. Now we sync on `writeBatchProtector` and after that `watchProcessorMutex`, we must preserve this order.

Reply:
These are never nested into each other